### ISSUE TITLE
Make it easy to use C# for Windows plugins

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#64958]
The primary language for Windows plugins is C++, since we don't want to force developers to take on the managed language overhead if they don't want to, but we should support C# since it's widely used in modern Windows development.

There are at least two levels to this:
- Minimal: Get loading and calling into a C# DLL working (it [currently doesn't](https://github.com/google/flutter-desktop-embedding/issues/531#issuecomment-654443486), and make a plugin template option that has that pre-configured. In this model, all the actual channel logic (extracting arguments, building responses, etc) still has to be written in C++, which is not ideal.
- Complete: Implement all the high-level channel constructs again in C#, and have the C++ code be a simple passthrough to C#. In this model, developers should never need to modify the C++ code.

**Comment by JakeSays** [comment]
@stuartmorgan I did some research on doing this a few months back as I have a client that wants C# plugins. Your "Complete" level is the route I decided to go. At the most basic level the native side would just host the CLR and marshal json (or maybe something more efficient) back and forth. All of the "C# plugin" management would be handled in C#, which is what I think you mean by channel constructs?

**Comment by stuartmorgan** [comment]
> Your "Complete" level is the route I decided to go. At the most basic level the native side would just host the CLR and marshal json (or maybe something more efficient) back and forth.

Assuming you mean writing a generic bridge between the C++ APIs and C#, that's a level between the two I listed actually.

> All of the "C# plugin" management would be handled in C#, which is what I think you mean by channel constructs?

By channel constructs, I mean most of the code in https://github.com/flutter/engine/tree/master/shell/platform/common/cpp/client_wrapper. In the Complete model the C++ client wrapper code isn't used at all.

**Comment by stuartmorgan** [comment]
> Also my target platform is actually linux, not windows.

Having the C# code necessary for "Complete" might be useful to you for that use case, but it's not one we're planning on explicitly supporting in Flutter.

**Comment by JakeSays** [comment]
> By channel constructs, I mean most of the code in https://github.com/flutter/engine/tree/master/shell/platform/common/cpp/client_wrapper. In the Complete model the C++ client wrapper code isn't used at all.

Ah ok - I haven't used the clent_wrapper yet (it didn't exist when I started writing plugins) so I'll take a look at that.

> Having the C# code necessary for "Complete" might be useful to you for that use case, but it's not one we're planning on explicitly supporting in Flutter.

Yeah all of the embedders I've written target "embedded" linux on small devices for use in full screen style applications. But as long as C++/cli wasn't used there's no reason why the same code wouldn't work on both platforms. The hosting API of .net core is very basic and easy to use from C++.

**Comment by shaxxx** [comment]
This issue is a great news. Any kind of support, even minimal, for calling .NET libraries from Flutter would be highly benifical.

Pardon my ignorance, but why would this be limited only for windows since .NET Core is cross platform? Is it because different binary formats between .net libs on windows and linux? Is this matter of "apples and bananas" or more like "if 90% of people prefer apples, we'll not sell any other fruit for now".? 

.NET 5 will be out soon and will merge Standard and Core version into single framework making it (mostly) cross platform.

Thank you for your work.

**Comment by JakeSays** [comment]
@shaxxx It's probably limited because the goal of flutter is to support plugins written in the host environment's most popular language, which is c# for windows, swift for ios/macos, and java for android. But there's nothing saying the code couldn't be generic enough to support any platform where core runs, but it probably wouldn't be supported by flutter on platforms other than windows.

Also .net5 isn't a merger of .net and core as most people think. It's mostly a rebranding of .net core 3.1 - ms was a little misleading.

**Comment by stuartmorgan** [comment]
> Pardon my ignorance, but why would this be limited only for windows since .NET Core is cross platform?

Because everything about the plugin system in Flutter is platform-specific.

> But there's nothing saying the code couldn't be generic enough to support any platform where core runs

Every embedding has its own plugin API, so this is not correct. A C# wrapper around the Windows plugin APIs will only work with the Windows embedding.

**Comment by JakeSays** [comment]
@stuartmorgan Why would the c# wrapper wrap the windows plugin APIs? Seems like it would make more sense to make a generic c# plugin that can load a platform specific assembly. This way each platform could have its own c# platform plugin. C#/.net is popular on more than just windows, and linux/osx are first class citizens. I am not suggesting the platform assemblies for linux/osx be written/supported by flutter, just that the ability exists without having to write yet another c# plugin.

**Comment by stuartmorgan** [comment]
> Why would the c# wrapper wrap the windows plugin APIs?

I don't understand what this question means. The only way to write a plugin for a Flutter embedding is against that embedding's plugin API.

> Seems like it would make more sense to make a generic c# plugin

I don't know what "a generic C# plugin" means. There is no such thing as a "generic" plugin for Flutter.

> that can load a platform specific assembly.

I don't know what this means either. It's not clear to me at this point whether you are using the term "plugin" to refer to the same thing this issue is about.

**Comment by JakeSays** [comment]
@stuartmorgan 

1 - I might be missing something but isn't the c++ wrapper the plugin extension point? If so then does each desktop platform also use the c++ wrapper? If that is also the case then there could be a "generic" c++ plugin (using the c++ wrapper) that hosts the CLR and supports implementing c# plugins. This should be fairly generic across the desktop platforms.
2 - See 1
3 - I am, but after thinking about it there isn't much of a need for a platform specific c# assembly - that's up to the plugin implementer.

I should note that I'm only talking about the desktop embeddings, although a custom embedder could copy/use the code in a similar fashion.

**Comment by shaxxx** [comment]
@stuartmorgan Thank you for your time. 
As far I can understand platform plugins is layered as

1.  Flutter (dart part) that calls methods via platform.invokeMethod
2.  Flutter (CPP part) that serializes all method calls and parameter and passes them to client wrapper
3.  Client wrapper (link provided here) that wraps implementation of MethodChannels for each platform and passes them to platform specific flutter engine
4.  CPP implementation of platform specific code (Flutter engine referenced for Android, iOS, Windows project)
5. Native plugin written in language specific to platform (Kotlin, Swift, C++, and in this proposal C#)

For second proposal (more complete) to be able to use C# in number 5, number 3 has to be implemented in C# to call 4 under the hood? This way native plugin wouldn't know anything about C++ part.

For first proposal this would require some working inside number 4 to allow calling C# libs from C++?

**Comment by JakeSays** [comment]
@shaxxx #3 would be the native CLR host which would load the c# plugins directly.

**Comment by stuartmorgan** [comment]
> 1 - I might be missing something but isn't the c++ wrapper the plugin extension point?

No. The C++ code is a convenience utility for use in plugins, so that they don't have to be written against the C API.

> If so then does each desktop platform also use the c++ wrapper?

No.

**Comment by EbramTawfik** [comment]
I think adding the "Complete" support for .net core will be very helpful writing **many** plugins for embedded and desktop development: Windows/Linux/macOS. Even if plugin system in Flutter is platform-specific you can write the shared code in .net core and compile and release it for each platform separately.

**Comment by stuartmorgan** [comment]
Again, this issue is only about Windows, and the work to write a C# bridge to the Windows plugin APIs *will not transfer* to other platforms.

**Comment by JakeSays** [comment]
How unfortunate.

**Comment by jtoy** [comment]
not using C# for windows means that flutter will not really be used on windows. What is the reasoning behind this?

**Comment by jhancock4d** [comment]
1. The argument for win32 versus UWP seems to be backwards compatibility. Except the only OS that doesn't support UWP on windows is Windows 7/8. Both of which are EOL and thus not supported and market share (legally) is approaching 0.  No new project would target Windows 7/8 because it is EOL. And any existing project that would add flutter embedded into an existing windows application can simply use XAML islands to load the UWP code. So there is absolutely no case where using Win32 is preferable to UWP from a pure compatibility standpoint.

2. Conversely, win32 doesn't run on Xbox (prime candidate for Flutter), HoloLens, nor the upcoming Windows 10x (win32 containers are delayed until 2022).  And even when Windows 10x supports win32 it will do so through containers that expose RDP endpoints so the UI execution will be significantly slower as is intended only for Line of Business applications, not for games nor other media/animation intensive applications -- i.e. flutter apps.  

UWP runs on ALL Windows environments. So from a "reach" perspective there is no reason to support win32 at all either because win32 now has less reach on windows than UWP does. (and Microsoft's guidance is all new apps that aren't system utilities should use UWP for this very reason and WinUI exists only to support legacy software that will never be ported to UWP.) (and those system utilities won't run on Windows 10x either)

3. Win32 code in C/C++ is VASTLY more difficult to write. Hence flutter targeting Win32 will always have a dearth of plugin support especially from 3rd parties.  If it was win32 that also supported .NET then you'd have more, but you still wouldn't have the quality of UWP with C/C++ OR C# plugins.  A quick look at job boards and as someone that hires and manages developers, Win32 C++ coders cost 40% more than a C# developer. And there are WAY fewer of them. All of that conspires to ensure that plugins won't be ported to windows where they are on other platforms, and if the plugins don't exist, then end application developers have to either pay/hire to extend them to support windows, or not use flutter, or not target windows. (making flutter on windows pointless)

**Example:** A video player with DRM and captioning as part of a plug-in would take 5 lines of c# uwp code.  It would support everything trivially including all of the things that are currently missing from the flutter video player (DRM, Captioning including WebVTT, full screen, playback speed, Http Headers, etc.) with very little effort. So anyone including google creating a video player that supports windows would be able to do so trivially and quickly and with expertise that is very widely available and about 40% cheaper than a win32 C++ developer.  (I know, I've done this very thing hence my example, but it's true of every platform interface)

The equivalent video player with the same codec, captioning, DRM etc. support would be over 5000 lines of code in win32. (or would have the equivalent dependencies. It would cost about 300% more to get done than the UWP + C# version which would take hours versus weeks or months with the C++/win32 version.

4. UWP + C# code is natively compiled and runs at over 99% of the speed of C++ code and is no larger in final memory or runtime than the C++ code.

5. UWP + C# code is VASTLY more secure and memory safe than C++ code, so the quality of the plugins that you'd get is vastly better.

6. UWP + C# is natively compiled and AOT. That means it runs at 99.4% on average the speed of C++ doing the same thing and similar memory usage as well and the final size on disk is similar.

So by targeting win32 you ensure that flutter adoption for windows apps will be low because the plugin availability will be low and the cost of adding the support that most apps need for plugins will be VERY high and you get no benefit for compatibility nor speed from win32 + C++ because there is no OS currently supported on windows that doesn't also support UWP and there are many platforms that do support UWP than don't support win32 and that list is about to get much larger this spring.

By targeting UWP those plugins can be written in C++ OR C# OR F# AND you get the entire UWP platform space which has all of the modern functionality easily available, so you'll get more adoption because the plugins will be easily ported to windows, and thus apps will be easily written to support windows without having to write more platform code and even if you do need to write platform code yourself, you'll be able to write vastly easier, and cheaper code to do so.

From what I can gather from this request, since it was written, WinUI has come along which is a standardization of the UWP controls and access to the UWP SDK for legacy applications. Microsoft explicitly says that new applications that aren't system utilities should still be written in UWP + the UWP version of WinUI and NOT WPF or Win32 + WinUI.  But that this functionality allows those legacy apps to access the UWP SDK (i.e. encrypted video players, biometrics, GPS, bluetooth etc.) directly in the same way that UWP apps can work.

So in theory, Flutter could support WinUI and get there. But the question is why? Because even if it was win32 + WinUI the overhead is the same as UWP apps AND it still couldn't target xbox, hololens or windows 10x.  And without .NET wrappers for that win32 hooks, you're still forced to write C++ code which is a far more specialized space specially in the windows UI space and thus more expensive to do for no actual benefit.

Hence as the author stated, there is no math that anyone has explained on the flutter team, that supports expending resources on Flutter win32 in any shape or form including in parallel to a UWP version which it is my understanding as a POC and is even working on Xbox.  There literally is no use case for a win32 version of flutter.  In every case the UWP version would be significantly better suited AND will encourage more adoption of flutter on windows because of the platform code problem.

**Comment by JakeSays** [comment]
@jhancock4d Where do you get 5k lines vs 5 lines? What makes UWP simpler in this case?

**Comment by jhancock4d** [comment]
@JakeSays 

In C# UWP (or C++ UWP) you'd just use media element and media player and attach to 3 events that are fired to handle AES, DRM, and HTTP Headers. And then Playback speed, full screen etc are simply methods on the media player.  It's actually 15 lines of code, not 5. But you get the point. It's all built in and just works. Same for bluetooth, same for GPS, same for GPIO, same for biometrics, etc. etc. etc. You have to write no more code. And every UWP app can support essentially all codecs on the market today either directly or by the user installing a UWP plugin codec from the store that microsoft provides. (i.e. HVEC, raw, ATOMOS, etc. are all free in the store and you can request they be installed automatically in your code and it will go get them for you)

On Win32, you have to create an hwnd, then import all of the stuff that replaced DirectDraw, import all of the playready code for DRM, add code to render text on the video after you draw the video to the hwnd, add code for http requests and intercept the http request for the m3u playlist and request it with the header information, get it back, parse it, add the caption information to the m3u file, then pass it to the video playback win32 calls and then do the same with playready.

After you wrote all of that and made sure that all of that code (which is several hundred lines of code) you'd have a player that would do DRM, Captions, full screen. HTTP Headers, and playback speed. BUT.

It wouldn't support anything other than wmv and mpegts as transports nor would it support h264/h265, nor anything other than wav audio. To add those you'd either have to pay several thousand dollars for other people's work or write it yourself against the arcane interfaces for creating your own container, codec and audio decoding.  

All in, it's about 5000 lines of custom code or about $15,000 in 3rd party libraries + your own code which adds up to about 5000 lines of code.

If you were to do the same in Win32 + WinUI + C++ code you'd have C++ code that would look like C++ version of the UWP + C# code above.  But you'd still have to write C++ code which is a tiny fraction of developers on windows AND the resulting app wouldn't run on Xbox, HoloLens, nor Windows 10x.

So what exactly do you win by using Win32 + C++ for flutter? (Skia runs about the same on UWP as it does on win32)

I sure can't see it. Which is why, until there is a fully supported version of Flutter for UWP we won't be investing the time in windows + flutter. There's just no upside. It's bad enough the number of plugins we've had to contribute to for Web, it will be WAY worse for Windows + win32.

**Comment by JakeSays** [comment]
Well, personally I think including the dependencies in to your 5k lines of code isn't valid. If you do that then you need to include the windows dependencies in the uwp case.

That aside, I think you are over-simplifying things. As it stands rendering video inside flutter is difficult and requires a lot of c++, regardless of the platform (except for ios/android, of course). It is my understanding that the plugin framework does not expose the rendering pipeline to the plugin, however this may have changed.

I know that to render video in the embedding framework requires the plugin to render each frame to a FBO/texture and then have flutter render the result.

**Comment by JakeSays** [comment]
@jtoy I'm not sure what you're getting at - this issue is about adding c# plugin support.

**Comment by jhancock4d** [comment]
@JakeSays  The UWP framework code is included in all versions of windows. It isn't an extra set of code that you have to deploy with your app. It's also freely usable for all Windows apps. The C++ dependencies are both not freely distributable for plugins (you'd have to pay even more for that) AND they would add to the deploy size of your flutter app. If you write it yourself, you still have to deploy it with your app bloating the size of the download. (and since UWP can be shared, memory used comparatively too because UWP is almost certainly in memory already and your code is not)

And UWP works EXACTLY the same as iOS and Android so plugins would be written exactly the same as iOS and Android NOT like Linux or win32.  That means that you would simply supply the control for the UWP plugin like you would with iOS and Android in SWIFT and Kotlin/Java.  Win32 would have to be able to create an hwnd for video to work at all.

If you look at the android code for the video player: https://github.com/flutter/plugins/blob/master/packages/video_player/video_player/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java you'll note that it isn't doing any of what you're suggesting.

Same with iOS.  And the UWP app would have even less code than the Android one. The win32 one would be horrific exactly as you're suggesting.

So yes, this issue should just be "Add UWP windows and only UWP windows so that resources are diverted to win32 which won't ever be up to par with iOS/Android".

**Comment by jtoy** [comment]
> @jtoy I'm not sure what you're getting at - this issue is about adding c# plugin support.
My bad, I meant to say that I don't believe using C++ makes sense for flutter and that instead it should be C#.

**Comment by JakeSays** [comment]
@jhancock4d actually that java code is doing exactly what i'm talking about. The native android player is rendering directly to a texture which is attached to a surface that is integrated in to the flutter rendering pipeline by the native android code: [https://github.com/flutter/engine/tree/master/shell/platform/android](https://github.com/flutter/engine/tree/master/shell/platform/android). It isn't as simple as just 'supplying a control'.

That is the level of integration that is needed in order for video to be seamlessly rendered by a UWP app, and it's a lot of c++ code. I would hope that the flutter guys would do that work when they do the uwp port, but who knows.

I've had to do this exact thing for other flutter ports and it's not trivial.

**Comment by JakeSays** [comment]
@jtoy actually c++ makes a lot of sense, just as much as c#.

**Comment by jhancock4d** [comment]
@JakeSays  Right, but that is done once by the Flutter team. Not by the developers of plugins.  Which is the point. Everything I originally said is just reinforced by your follow-on comments, not weakened.

**Comment by JakeSays** [comment]
@jhancock4d Well, I think we're coming at this from different angles. I am talking about the amount of work the flutter team needs to do, and you're talking about the developer experience. Of course the developer experience would be the same across all plugins if the flutter team did the work.

There are many shops that have huge investments in c++, both in IP and talent, and there is no cost benefit in rewriting everything in c#. The more languages supported the better. So what may not make sense for you makes a ton of sense for others. You seem to make the assumption that all future work is either green field or would be better off if rewritten.

**Comment by JakeSays** [comment]
Actually the uwp port is significantly more difficult than other platforms, including win32, as uwp doesn't support opengl. The flutter engine uses ANGLE which is a huge dependency.

**Comment by JakeSays** [comment]
@stuartmorgan Quick question for you. For plugins we have swift for ios, java/kotlin for android, and c# for windows. Is there a story for linux desktop plugins? As far as I know there isn't one "defacto platform language" for linux like there is for the other platforms.

**Comment by jhancock4d** [comment]
@JakeSays 

1. It is vastly better for the flutter team to put in the work ONCE to write the UWP version and have plugins be VASTLY easier to write after EXACTLY as they did for iOS and android. (in both cases they could have make plugins just as difficult for Windows as the win32 version will be)

2. Those shops that have investments in C++ can use C++ to write UWP plugins. There is no one stopping them.
3. Those shops that have C++ win32 apps (very, very, very rare, over 95% of all apps made for windows do not use c++/win32 or even MFC) can use XAML islands to load flutter inside of them with the same ease as it would take to do it with a win32 implementation of flutter.
4. Actually there is already a layer for ANGLE on UWP provided by Microsoft AND a newer version that is provided by google. Hence the dependency is already met as evidenced by SKIA running just fine on UWP. (ultimately UWP still uses hwnds that you can write to so there nothing stopping you from hooking an opengl package on it.
5. There already is a port of Flutter for UWP: https://www.windowslatest.com/2020/09/21/google-flutter-uwp-apps-for-windows-10/
6. Linux uses GTK+ I believe.

So again, you're proving my point. If you want to write C++ code on UWP, go ahead. I don't know why you'd do so, but you're free to do so. 

Having the flutter team invest the time in doing the  coding for the flutter renderer for UWP is vastly better than having every single plugin have to invest similar time which is what would happen if anyone actually bothered which we know for sure they won't. Just look at how few modern windows apps there are. Why? Because the effort to write a windows app isn't justified when there is a web browser on there that works fine for most apps and ios and android own the rest.  The ONLY reason that flutter on windows will ever be used is if it's low effort to do so. I.e. deploy your code to the store and it just works.  This is the same reason why Apple is bringing iOS apps to Mac. Why would you invest in a mac app when there is a perfectly good web browser right there and most everyone else is on iOS or Android?  The answer is that devs weren't. So apple made iOS apps easy to run on MacOS and because it takes little effort there's a massive influx of apps on Mac already.

Your assertions don't reflect the marketplace nor the reality of Windows development.  There is nothing stopping legacy win32 apps from using UWP. There's nothing stopping C++ developers from using UWP. But NO win32 app will work on Xbox, Hololens, or Windows 10x (this spring) and less than 5% of windows developers will be able to contribute to flutter plugins ensuring that they just won't get written which ensures that flutter for windows won't be used.

Honestly, flutter win32 is a waste of resources. If Google isn't willing to do UWP, they're just throwing away money creating Flutter win32 because adoption will be approaching 0.

**Comment by JakeSays** [comment]
Yes I am aware there is a UWP port of flutter. I've known that all along. And yes I am also aware that uwp uses an hwnd, and that there's an abandoned ms port of angle, and that google also has a port (which makes sense since angle is a google product). The work needed to get angle to work on uwp for flutter was done by the flutter team.

I guess I don't understand what your issue is. Flutter is a google product and they're approaching it from the standpoint that makes sense for them.  They have already committed to a UWP port.

And linux uses gtk+/qt/wx/raw X/and dozens of others.

**Comment by jhancock4d** [comment]
@JakeSays The original poster's issue is that at the time there was no plans for UWP at all.

My issue now is that resources are being expended for a win32 version that has no business case and those resources could be better used to make the UWP version better and be brought up to snuff or expand the functionality of plugins that are managed by google (like the anemic video player for example)

So as a relatively large customer that needs Windows and doesn't want to have to maintain a separate app because of the decisions of the flutter team, we have a vested interest.  We can say for absolute certainty that using flutter win32 won't work for us and would cost more than writing our apps in UWP in addition to flutter for everything else. We've done the cost analysis on it.

I personally consult with multi-national banks as well and every one of them has exactly the same issue (they will just punt and not bother with a windows app at all and just continue to push people to the web site).

**Comment by JakeSays** [comment]
@jhancock4d Ok but that's just your company. There are countless others with different needs that you are not aware of. Google may have significant business needs for a win32 version before uwp that they're just not talking about. Who knows.

Frankly I disagree with some of the flutter team's decisions but since I'm not forking over the scratch to get the work done, I'm not in a position to make demands. I just implement the missing functionality myself.

**Comment by jhancock4d** [comment]
@JakeSays My analysis is based on market metrics. It isn't my needs. I simply used myself and my clients as examples.  The math doesn't lie.

Google has < 10 apps on windows of any significance. (chrome, google drive, android studio, and...?)  They aren't going to be using flutter in Android studio (which is JetBrains anyhow) chrome doesn't need it for anything because it's a web browser (maybe settings ui but why have the overhead when the rest is native win32?), and google drive might use it for UI but nothing that I'm outlining above would affect google drive's ability to use flutter since it would just use Xaml islands for the UI if they really wanted to. Ya, windows 7 support would go away, but so what? By the time they actually released it fewer people will use windows 7 than Linux on the desktop.

If google decided to not smother windows as they have in the past especially on mobile and released YouTube and other apps, all of those are green field and no reason why they couldn't be in UWP and using UWP would significantly benefit that team because plugin development would be vastly easier.

So unless someone in Google thinks that there's value in supporting an OS that is 14 years old, and virtually no one uses, there is no use case for google.

So back to my point.

**Comment by britannio** [comment]
This issue is tracking "Make it easy to use C# for Windows plugins"
Progress is being made in regards to supporting UWP https://github.com/flutter/flutter/issues/14967

**Comment by JakeSays** [comment]
You're being myopic. I personally know of a company with several thousand kiosk displays running windows embedded. They're looking at moving to flutter for the apps because it fits well with some of their other products. This 14 year old OS that you mention is working perfectly fine. The business cases I mention might not be Google products.

But this conversation is going in circles. Believe what you will - to me it's irrelevant.

**Comment by jhancock4d** [comment]
@JakeSays Within the next 2 years that company will be forced to move because Microsoft won't even let you pay for patches. Or they'll be one of those horrific security disasters like we had with Windows 2000 and XP embedded with banks.

So by the time flutter for windows (whatever it is) is ready for production code for those kiosks to run it, they won't be able to run Windows 7 or anything that doesn't support UWP in a production environment because their business insurance won't let them. (we already are asked and our policies go up unless we have a support agreement for windows 7 patches from Microsoft and every 3 months those support agreements increase in cost)

And in that environment they'll almost certainly have to green field develop so they won't even have to use Xaml islands to use flutter on windows anyhow.

So again, you're searching for justifications for something that makes no sense. (Which is why this is going around in circles.)

**Comment by JakeSays** [comment]
@jhancock4d lol you keep making assertions about things you know nothing about, nor could you know because it's not your place to know. In two years, when the time is right, this company will slide effortlessly over to linux because of the work we're doing now using flutter ON win32. Other than the win32 embedder I wrote none of it is greenfield. It would have been insanely stupid to have done anything with uwp, and there is huge value in starting with win32.

My point has been all along this: Your situation is not the entirety of all situations in the world.

**Comment by jhancock4d** [comment]
@JakeSays If you slide over effortlessly to Linux then every one of those plugins you use on windows will also have to be on linux.  Your win32 plugin code will not port any easier than UWP plugin code would if you wrote said plugins in C++ with UWP. And in most cases those plugins will have to be written from scratch anyhow because it's so platform specific, especially with linux that has no standard way of doing virtually anything in UI and things that aren't UI will have no consistency at all.

But at least now we know why you're so vehement about this: You have a vested interest and are the small exception to the data (i.e. less than 5%).

Would have been nice if you'd disclosed that from the beginning as it would have saved everyone time in the conversation.

Yes I'll stipulate that in your incredibly specialized case, this *might* be slightly more difficult to go over to Linux later. And even then that's *might* given that your win32 enabler could easily just create a xaml island for the flutter app to run inside of which is no more complex.

But your incredibly specialized case is exactly that. It's an edge case that doesn't create a market.  The VAST majority of people will write embedded versions in flutter green field OR they'll stay on Windows Embedded and add flutter for their migration to Windows 10.

So again, you're proving my point: 95+% of use cases are better served with UWP than win32 AND because of that and the lower overhead to writing plugins for windows in UWP, flutter will get vastly better adoption on windows as a UWP app than on win32. I'd guess based on all of the UWP activity and very little win32 activity in comparison that Google has figured this out too.

Over and out.

**Comment by stuartmorgan** [comment]
This issue is not about UWP vs Win32. Repsting arguments you've already made in the UWP issue on an issue where they are off topic is unproductive.

**Comment by jhancock4d** [comment]
@stuartmorgan Then what's the use case for .net with a win32 implementation?  Are you requesting the ability to use winforms objects for things like video players etc?  .NET without UWP would have limited use because you don't have access to the API and if you do have access to the api (winui) then you'd be better served with UWP flutter with XAML islands in the original code because of the much richer UWP stack because you're already on windows 10.

And using .net without UWP incurs a massive penalty of .net runtime overhead even if you do a full native compile.

So please let us know what your intent is because I don't see the use case in the data without UWP.

**Comment by stuartmorgan** [comment]
You've made it abundantly clear that you don't see any value in Win32 support; I see no productive outcome to having yet another version of that discussion in the context of C# support.

**Comment by yanshouwang** [comment]
Is there any way to develope flutter plugin for windows with C# now? I'm familiar with C#, Kotlin, Swift, but C++ is really hard to me...

**Comment by Sunbreak** [comment]
> Is there any way to develope flutter plugin for windows with C# now? I'm familiar with C#, Kotlin, Swift, but C++ is really hard to me...

https://github.com/timsneath/win32 and dart:ffi may be a better choice than C++

**Comment by jhancock4d** [comment]
@Sunbreak That's excellent and makes this viable.  What would be really interesting is a POC of using project reunion libraries. This would enable video playback with DRM, OpenIdConnect authentication windows, and more without having to use UWP.  Are you aware of any efforts on this front?

**Comment by stuartmorgan** [comment]
Again, this issue is specifically tracking enabling the easy use of C# when developing Windows plugins. It is not a general discussion forum for all topics related to Windows plugin development.

If you want to discuss FFI-based development, please do so in another forum.

**Comment by ManasMalla** [comment]
> @jtoy actually c++ makes a lot of sense, just as much as c#.

> @stuartmorgan Then what's the use case for .net with a win32 implementation? Are you requesting the ability to use winforms objects for things like video players etc? .NET without UWP would have limited use because you don't have access to the API and if you do have access to the api (winui) then you'd be better served with UWP flutter with XAML islands in the original code because of the much richer UWP stack because you're already on windows 10.
> 
> And using .net without UWP incurs a massive penalty of .net runtime overhead even if you do a full native compile.
> 
> So please let us know what your intent is because I don't see the use case in the data without UWP.


Actually in the times like these where C# is given more support for most of the libraries by the official Microsoft Documentation, I guess the UWP api are given backward compatibility to win32 apps too because when I was trying out the implementation of a plugin for Windows on a regular VS project then the app ran both on WIN32 and WINUWP using the same API's I implemented which were documented for UWP in C#. So I guess the implementation for C# to write plugins for flutter would be most beneficial as it would allow the enthusiastic developer community to quickly port the most of the Windows APIs to flutter giving opening up a new breakthrough for the world to develop apps! I am most enthusiastic to begin in this work and hope all of you would join me in doing the same!  I guess we would need to begin with the step mentioned by @stuartmorgan in this post - 





> @stuartmorgan Thank you for your time.
> As far I can understand platform plugins is layered as
> 
> 1. Flutter (dart part) that calls methods via platform.invokeMethod
> 2. Flutter (CPP part) that serializes all method calls and parameter and passes them to client wrapper
> 3. Client wrapper (link provided here) that wraps implementation of MethodChannels for each platform and passes them to platform specific flutter engine
> 4. CPP implementation of platform specific code (Flutter engine referenced for Android, iOS, Windows project)
> 5. Native plugin written in language specific to platform (Kotlin, Swift, C++, and in this proposal C#)
> 
> For second proposal (more complete) to be able to use C# in number 5, number 3 has to be implemented in C# to call 4 under the hood? This way native plugin wouldn't know anything about C++ part.
> 
> For first proposal this would require some working inside number 4 to allow calling C# libs from C++?

**Comment by datocrats-org** [comment]
Brainstorming on how to approach this, I think Microsoft's page shows their current strategy, worth reading to anticipate how to do this.  They will write from C# to C++ soon they say, will that help to do the reverse? https://blogs.windows.com/windowsdeveloper/2021/01/21/making-win32-apis-more-accessible-to-more-languages/

**Comment by stuartmorgan** [comment]
No brainstorming is needed at this point. The first step here is still:

> * Get loading and calling into a C# DLL working (it [currently doesn't](https://github.com/google/flutter-desktop-embedding/issues/531#issuecomment-654443486)), and make a plugin template option that has that pre-configured.

Almost all of the discussion here since the issue was filed is mostly off-topic, which may make this seem less well-understood than it is; the first step is just for someone to build a working CMake version of [this](https://github.com/google/flutter-desktop-embedding/issues/531#issuecomment-608145389).

> They will write from C# to C++ soon they say, will that help to do the reverse?

Language projections aren't bi-directionnel, so no. But the API surface here is small, so a generalized language protection isn't necessary.

**Comment by yanshouwang** [comment]
Any progress? this is the key issue for our team to develope windows apps with flutter, we use C# a lot, but C++ is not easy for everyone

**Comment by urusai88** [comment]
We need support for .NET based flutter apps!

**Comment by selvam920** [comment]
+1

**Comment by LiveOrNot** [comment]
> Any progress? this is the key issue for our team to develope windows apps with flutter, we use C# a lot, but C++ is not easy for everyone

https://www.cnblogs.com/liveornot/p/16705613.html

**Comment by LiveOrNot** [comment]
[> > @jtoy实际上 c++ 很有意义，就像 c# 一样。
> 
> > @stuartmorgan那么带有win32实现的.net的用例是什么？您是否要求能够将 winforms 对象用于视频播放器等？没有 UWP 的 .NET 的使用会受到限制，因为您无权访问 API，并且如果您确实有权访问 api (WINUI)，那么在原始代码中使用带有 XAML 岛的 UWP flutter 会更好，因为更丰富的 UWP 堆栈，因为您已经在 Windows 10 上。
> > 即使您进行完整的本机编译，使用没有 UWP 的 .net 也会导致 .net 运行时开销的巨大损失。
> > 所以请让我们知道您的意图是什么，因为我在没有 UWP 的数据中看不到用例。
> 
> **实际上，在这样的时代，官方 Microsoft 文档**为 C# 提供了对大多数库的更多支持，我猜**UWP API 也被赋予了对 win32 应用程序的向后兼容性，**因为当我尝试在 Windows 上实现插件时一个常规的 VS 项目，然后该应用程序使用我实现**的相同 API 在 WIN32 和 WINUWP 上运行，这些 API****在 C# 中为 UWP 记录**。此外，大多数 API 不适用于 C++，因此使我们无法开辟新的领域。我什至想提一下，安装包的 NuGet 系统是由@bshiva1987
> 
> > https://github.com/bshiva1987/flutter_win_webview/blob/master/webview_popupauth/windows/CMakeLists.txt具有进行手动 NuGet 包安装的概念验证，直到这里有解决方案。
> 
> 所以我想**C# 为 Flutter 编写插件的实现将是最有益**的，因为它可以让热情的开发者社区快速将大部分 Windows API 移植到 Flutter 中，从而为世界开发应用程序开辟新的突破！**我最热衷于开始这项工作，希望你们都和我一起做同样的事情！** 我想我们需要从提到的步骤开始@stuartmorgan在这篇文章中 -
> 
> > @stuartmorgan感谢您的时间。
> > 据我所知，平台插件是分层的
> > 
> > 1. ****通过 platform.invokeMethod 调用方法的**Flutter （飞镖部分）**
> > 2. **Flutter **（CPP 部分）序列化所有方法调用和参数并将它们传递给客户端包装器
> > 3. **客户端包装器**（此处提供链接），它为每个平台包装 MethodChannels 的实现并将它们传递给特定于平台的颤振引擎
> > 4. **平台特定代码的CPP 实现**（Android、iOS、Windows 项目引用的 Flutter 引擎）
> > 5. ****以特定于平台的语言编写的本**机插件（Kotlin、Swift、C++ 和本提案中的 C#）**
> > 
> > 对于第二个提案（更完整）能够在数字 5 中使用 C#，数字 3 必须在 C# 中实现才能在引擎盖下调用 4？这样原生插件就不会知道任何关于 C++ 部分的信息。
> 
> 我想我们可以坚持第二个建议，在保留 C++ 支持的同时完全添加对 C# 的支持！



> @stuartmorgan Thank you for your time. As far I can understand platform plugins is layered as
> 
> 1. Flutter (dart part) that calls methods via platform.invokeMethod
> 2. Flutter (CPP part) that serializes all method calls and parameter and passes them to client wrapper
> 3. Client wrapper (link provided here) that wraps implementation of MethodChannels for each platform and passes them to platform specific flutter engine
> 4. CPP implementation of platform specific code (Flutter engine referenced for Android, iOS, Windows project)
> 5. Native plugin written in language specific to platform (Kotlin, Swift, C++, and in this proposal C#)
> 
> For second proposal (more complete) to be able to use C# in number 5, number 3 has to be implemented in C# to call 4 under the hood? This way native plugin wouldn't know anything about C++ part.
> 
> For first proposal this would require some working inside number 4 to allow calling C# libs from C++?

you can try this 'https://github.com/LiveOrNot/FlutterSharp](https://github.com/LiveOrNot/FlutterSharp)

> Windows 插件的主要语言是 C++，因为我们不想强迫开发人员承担托管语言开销（如果他们不想），但我们应该支持 C#，因为它广泛用于现代 Windows 开发。
> 
> 这至少有两个层面：
> 
> * 最小：加载和调用 C# DLL 工作（[目前不](https://github.com/google/flutter-desktop-embedding/issues/531#issuecomment-654443486)工作），并制作一个预先配置好的插件模板选项。在这个模型中，所有实际的通道逻辑（提取参数、构建响应等）仍然必须用 C++ 编写，这并不理想。
> * 完成：在 C# 中再次实现所有高级通道构造，并使 C 代码成为 C# 的简单传递。在此模型中，开发人员永远不需要修改非 C# 代码。



> The primary language for Windows plugins is C++, since we don't want to force developers to take on the managed language overhead if they don't want to, but we should support C# since it's widely used in modern Windows development.
> 
> There are at least two levels to this:
> 
> * Minimal: Get loading and calling into a C# DLL working (it [currently doesn't](https://github.com/google/flutter-desktop-embedding/issues/531#issuecomment-654443486)), and make a plugin template option that has that pre-configured. In this model, all the actual channel logic (extracting arguments, building responses, etc) still has to be written in C++, which is not ideal.
> * Complete: Implement all the high-level channel constructs again in C#, and have the C code be a simple passthrough to C#. In this model, developers should never need to modify non-C# code.

https://github.com/LiveOrNot/FlutterSharp (although this was my simple demo, it showed me that the P/Invoke hook flutter desktop makes it possible to develop flutter plug-ins using C#, looking forward to a better solution)

**Comment by loic-sharma** [comment]
@LiveOrNot Great work! Some thoughts:

1. C# is a managed language and will likely impact the app's performance. We should measure the impact on: 
    1. Compile time
    1. Startup time 
    1. Memory/CPU consumption
    1. App size, especially for .NET Core self-contained apps 
1. While .NET Framework is popular for existing apps, it is "completed" and not receiving feature updates. New applications should use .NET 7. Flutter should use .NET Standard to allow both scenarios.
1. Ideally a C# app would avoid assembly scanning & reflection for plugins. The Flutter tool should generate the plugin registration code. That said, your approach is perfectly reasonable for an MVP

**Comment by JakeSays** [comment]
@loic-sharma actually I disagree with #3 - the option should be left up to the plugin developer. We're big boys, let us choose how plugins should be loaded.

**Comment by JakeSays** [comment]
@LiveOrNot I looked at your FlutterSharp and it looks good, but it seems to be more of an "add to app" scenario where the flutter engine is "hosted" by a WPF application. I think this is a valid use case. Another use case is where .net is "hosted" by a flutter application. This is especially useful for plugins that do not have user interface requirements.

I have a prototype of using c# plugins from a flutter application on linux. I'll post it this weekend. It would be nice to support both scenarios.

**Comment by stuartmorgan** [comment]
> the option should be left up to the plugin developer.

Plugin loading is a function of the application, not the plugin. Plugin developers can't choose how the hosting system works.

> We're big boys, let us choose how plugins should be loaded.

You're certainly welcome to do whatever you like in your application, including building your own plugin system with whatever loading behavior you want. Flutter's plugin system, for loading Flutter plugins, is based on build-time registration.

**Comment by JakeSays** [comment]
> Plugin loading is a function of the application, not the plugin. Plugin developers can't choose how the hosting system works.

Yes, sorry I meant application developer, not plugin developer.



> You're certainly welcome to do whatever you like in your application, including building your own plugin system with whatever loading behavior you want. Flutter's plugin system, for loading Flutter plugins, is based on build-time registration.

Whatever.

**Comment by LiveOrNot** [comment]
> @LiveOrNot Great work! Some thoughts:
> 
> 1. C# is a managed language and will likely impact the app's performance. We should measure the impact on:
>    
>    1. Compile time
>    2. Startup time
>    3. Memory/CPU consumption
>    4. App size, especially for .NET Core self-contained apps
> 2. While .NET Framework is popular for existing apps, it is "completed" and not receiving feature updates. New applications should use .NET 6+. Flutter should use .NET Standard to enable both legacy and modern scenarios.
> 3. Ideally a C# app would avoid assembly scanning & reflection for plugins. The Flutter tool should generate the plugin registration code. That said, your approach is perfectly reasonable for an MVP
> 
> I'll play around with this! :)

In fact, I found that it is not the C # host program that occupies a lot of memory, but the Flutter itself. The Flutter engine will occupy about 90MB of memory, so I need to consider how to reuse the Flutter engine to host multiple windows; In addition, you can use the same method to host Flutter desktop applications in netcore, net5.0, and net6.0 desktop applications. But I have to say that there is a risk of memory leakage when using WPF to host Flutter Desktop applications, which I need to improve

**Comment by LiveOrNot** [comment]
> @LiveOrNot I looked at your FlutterSharp and it looks good, but it seems to be more of an "add to app" scenario where the flutter engine is "hosted" by a WPF application. I think this is a valid use case. Another use case is where .net is "hosted" by a flutter application. This is especially useful for plugins that do not have user interface requirements.
> 
> I have a prototype of using c# plugins from a flutter application on linux. I'll post it this weekend. It would be nice to support both scenarios.

I'm looking forward to your work. In fact, my next step is to use the same method in MAUI or directly call the platform's native API to implement the C # hosted Flutter plug-in registration mechanism. I have also tried to directly call the C # library in the Flutter Desktop application plug-in by turning on the CMake CLR switch, but the System AccessViolationException throws； I have also tried to call directly through COM Component and directly create CLR Runtime Environment in C++, but these two methods are not elegant enough (ha ha, I don't have much CMake development experience, so I gave up this method

**Comment by JakeSays** [comment]
@LiveOrNot theoretically my approach should work on windows as well with very little change. Basically I turn the flutter app in to a .net core runtime host and those hosting APIs are the same on all .net platforms, iirc.

**Comment by LiveOrNot** [comment]
Maybe the Native AOT mechanism of NET 7.0 can solve the problem of directly calling C # libraries in Flutter Desktop applications. If this method is feasible, it will be more comfortable for C # programmers

**Comment by LiveOrNot** [comment]
> @LiveOrNot theoretically my approach should work on windows as well with very little change. Basically I turn the flutter app in to a .net core runtime host and those hosting APIs are the same on all .net platforms, iirc.

Sounds good, ha ha, looking forward to it

**Comment by JakeSays** [comment]
> Maybe the Native AOT mechanism of NET 7.0 can solve the problem of directly calling C # libraries in Flutter Desktop applications. If this method is feasible, it will be more comfortable for C # programmers

Actually I believe .net 7 will support exposing c# methods as C callable function symbols, which in theory would enable a dart FFI scenario.

**Comment by LiveOrNot** [comment]
> > Maybe the Native AOT mechanism of NET 7.0 can solve the problem of directly calling C # libraries in Flutter Desktop applications. If this method is feasible, it will be more comfortable for C # programmers
> 
> Actually I believe .net 7 will support exposing c# methods as C callable function symbols, which in theory would enable a dart FFI scenario.
Yes，I think so

**Comment by loic-sharma** [comment]
@JakeSays [This comment](https://github.com/flutter/flutter/issues/64958#issuecomment-1251737257) may seem unwelcoming to others. Please read our [code of conduct](https://github.com/flutter/flutter/blob/master/CODE_OF_CONDUCT.md). In particular:

> * Respect people, their identities, their culture, and their work.
> * Be kind. Be courteous. Be welcoming.
> * Listen. Consider and acknowledge people's points before responding.

**Comment by loic-sharma** [comment]
> @loic-sharma actually I disagree with # 3 - the option should be left up to the application developer.

What are your concerns with Flutter generating code to register plugins? There are drawbacks to using reflection:

1. Regresses performance
2. Impedes app trimming, resulting in bigger apps

**Comment by JakeSays** [comment]
> What are your concerns with Flutter generating code to register plugins? There are drawbacks to using reflection:

I am familiar with the drawbacks to reflection, just as I am familiar with the drawbacks to static registration. On a desktop we have a lot more horsepower at our disposal, which means we can develop larger, more dynamic applications. Both the issues you mentioned are much less of a concern in a desktop environment.

All I am asking for is the flexibility I have from pretty much every other desktop application framework - leave the trade-off decisions in my hands so that I can develop my application the way that suites my needs. 

And your comment about flutter not preventing me, well, that's only technically true. Practically, if I don't do it the exact flutter way, I lose tooling support, have to write my own desktop embedding and plugin system, etc. So to say "hey you are free to roll you own.." is, well, offensive to me and shows a lack of respect for my time and my customer's money.

**Comment by stuartmorgan** [comment]
> All I am asking for is the flexibility I have from pretty much every other desktop application framework - leave the trade-off decisions in my hands so that I can develop my application the way that suites my needs.

Plugin registration is tangential to this issue, so it would be helpful to move that discussion to a new issue. The best way to proceed here would be to file a new issues using the Proposal template, and describe the goals of the proposal. (So far it's not clear to us what use cases this request is intended to address.)

**Comment by sme-tuga001** [comment]
Hi, new .NET 8 version is out now. With AOT support, can this somehow help solve this issue?

**Comment by chipweinberger** [comment]
Has anyone gotten C# to work for plugins? (and can share an example?)

It should be possible even without official Flutter support I think?

**Comment by JakeSays** [comment]
@chipweinberger I did a proof of concept a few years ago that had c# plugins working, but it requires a custom embedders as the official embedders don't support dynamically loading plugins. Technically, though, it should be possible to spin up a .net instance using dart's FFI without needing a custom embedder. The newer versions of .net (7+) have better support for native-.net marshaling.

**Comment by loic-sharma** [comment]
> Has anyone gotten C# to work for plugins?

This is possible without a custom embedder. Each Flutter Windows plugin expose a C API for registration. For example, the URL launcher plugin exposes [`UrlLauncherWindowsRegisterWithRegistrar`](https://github.com/flutter/packages/blob/316bde0b3908ee4b6f6928476ec1b4ecba393e39/packages/url_launcher/url_launcher_windows/windows/include/url_launcher_windows/url_launcher_windows.h#L19-L20).

You'll need to:

1. Compile your Flutter Windows plugins as they are distributed as source files. I'd recommend:
    1. Create a C++ Flutter Windows app
    2. Add your desired plugins to this app
    3. Run `flutter build windows`. This compiles your app and its plugins
    4. You can find the compiled plugins here: `build/windows/<arch>/runner/Release/<plugin>.dll`
2. Update your C# app:
    1. Update the project to reference your compiled plugin DLLs. You'll want to copy these DLLs to the output directory
    2. Update your C# entry point to P/Invoke your plugins' registration function.

This is seriously far from the golden path. I wouldn't recommend this unless you have a deep understanding of what you're doing.

**Comment by listepo** [comment]
Any news for official support?

**Comment by listepo** [comment]
@stuartmorgan-g any news? Do you need help?

**Comment by stuartmorgan-g** [comment]
> @stuartmorgan-g any news?

Please see https://github.com/flutter/flutter/blob/master/docs/contributing/issue_hygiene/README.md#do-not-add-me-too-or-same-or-is-there-an-update-comments-to-issues-or-prs

> Do you need help?

I am not currently working on this issue. Anyone interested in doing so is certainly welcome to; the first step here should probably be a working proof of concept and/or a design document laying out the planned approach.

**Comment by loic-sharma** [comment]
I started a prototype here: https://github.com/loic-sharma/butter

It is hacky and requires significant polish, but it is a working proof-of-concept.

I don't think this is something we'd merge into Flutter itself. We'd likely keep this a separate tool, similar to how third-party Flutter platforms like Flutter Tizen work.

**Comment by listepo** [comment]
@loic-sharma  @stuartmorgan-g I think С# should be part of engine, because this is not a new platform. I think a C# application should be the default. C++ is complex enough to be used as a default language. I'm also experimenting with C# support, is there any documentation on adding a new platform?

**Comment by listepo** [comment]
It would be nice if Google sponsored the addition of new features; the plan has its own budget for each feature depending on its complexity.

**Comment by stuartmorgan-g** [comment]
> I think С# should be part of engine

We have no plans to use C# in the engine, but that's unrelated to this issue, as supporting C# in plugins should not require engine-level changes. If you have a proposal related to the engine rather than plugins, please file a new issue with proposal details.

> I think a C# application should be the default.

Changing the application template is also orthogonal, and would also need to be filed as a new issue.

> I'm also experimenting with C# support, is there any documentation on adding a new platform?

I'm not sure what you mean by "adding a new platform". If you are exploring writing an entirely new Windows embedding in C# (which would also not be within the scope of this issue), you could try the `#custom-embedders` channel in Discord.

**Comment by listepo** [comment]
@stuartmorgan-g  thanks
